packages/fs-detectors/test/unit.framework-detector-record.test.ts

```diff
@@ -0,0 +1,194 @@
+import frameworkList from '@vercel/frameworks';
+import { detectFrameworkRecord } from '../src';
+import VirtualFilesystem from './virtual-file-system';
+
+describe('detectFrameworkRecord', () => {
+  it('Do not detect anything', async () => {
+    const fs = new VirtualFilesystem({
+      'README.md': '# hi',
+      'api/cheese.js': 'export default (req, res) => res.end("cheese");',
+    });
+
+    const framework = await detectFrameworkRecord({ fs, frameworkList });
+    expect(framework?.slug).toBe(undefined);
+  });
+
+  it('Detects a framework record with a matchPackage detector', async () => {
+    const fs = new VirtualFilesystem({
+      'package.json': JSON.stringify({
+        dependencies: {
+          next: '9.0.0',
+        },
+      }),
+    });
+
+    const frameworkRecord = await detectFrameworkRecord({ fs, frameworkList });
+    if (!frameworkRecord) {
+      throw new Error(
+        '`frameworkRecord` was not detected, expected "nextjs" frameworks object'
+      );
+    }
+    expect(frameworkRecord.slug).toBe('nextjs');
+    expect(frameworkRecord.name).toBe('Next.js');
+    expect(frameworkRecord.detectedVersion).toBe('9.0.0');
+  });
+
+  it('Detects a framework record with a matchPackage detector with slashes', async () => {
+    const fs = new VirtualFilesystem({
+      'package.json': JSON.stringify({
+        dependencies: {
+          '@ionic/angular': '5.0.0',
+        },
+      }),
+    });
+
+    const frameworkRecord = await detectFrameworkRecord({ fs, frameworkList });
+    if (!frameworkRecord) {
+      throw new Error(
+        '`frameworkRecord` was not detected, expected "ionic-angular" frameworks object'
+      );
+    }
+    expect(frameworkRecord.slug).toBe('ionic-angular');
+    expect(frameworkRecord.detectedVersion).toBe('5.0.0');
+  });
+
+  it('Detect first framework version found', async () => {
+    const fs = new VirtualFilesystem({
+      'package.json': JSON.stringify({
+        dependencies: {
+          nuxt: '1.0.0',
+          nuxt3: '2.0.0',
+          'nuxt3-edge': '3.0.0',
+        },
+      }),
+    });
+
+    const framework = await detectFrameworkRecord({ fs, frameworkList });
+    expect(framework?.slug).toBe('nuxtjs');
+    expect(framework?.detectedVersion).toBe('1.0.0');
+  });
```

Comment:
When you say "first" does this mean _first_ dependency in `dependencies` or _first_ detector in the frameworks list? This is pretty gnarly edge case since if the user had  all of these actually listed in their package.json - they either have a bug in their project somewhere or shouldn't be relying on our auto detection logic.

Reply:
"First" means the first detected version based on the order of the detectors, which I now realize isn't exactly tested by this test setup.

I don't love this either. The alternative is to return no version if we find more than one. I think that would be fine. We could log that it happened, too.

Reply:
Or we could go back to creating a separate key that identifies the package we should use to pluck version number. Then, the detector order wouldn't matter.

Previous discussion: https://github.com/vercel/vercel/pull/9009#discussion_r1045090088

Reply:
We could use a team decision on the strategy to take with using detectors to determine version of detected frameworks. I've implemented it different ways and there's not yet a consensus on the path to take.

Options:

- **(1)** create new property on framework entries like `versionDependency: 'nuxt3'` at the top-level of a framework entry
    - **(1a)** allow only a single package to be listed; if you want multiple, create multiple framework entries
    - **(1b)** allow an array of packages, preferring the first matched package to return version
    - **pro:** clearer place to determine version 
    - **con:** could be confusing (no matter what we name it) next to the existing (and mostly deprecated) `dependency` property
- **(2)** leverage detectors that already look at package.json dependencies like `matchPackage: 'nuxt3'` in a detector
    - **(2a)** if multiple detectors match, use the first version number
    - **(2b)** if multiple detectors match, use none of the version numbers
    - **(2c)** if multiple detectors look at package.json dependencies, throw error saying to create multiple framework entries instead
    - **pro:** we already scan package.json dependencies for specific packages; plucking out version can be done in the same action
    - **con:** less clear that this is happening

*Note:* There are cases where our framework config is the same for different packages that could match, such as `nuxt`/`nuxt3`/`nuxt-edge`.

Having implemented both options, I think option 1b offers the best balance of flexibility and clarity.

Reply:
I would vote for `(2a) if multiple detectors match, use the first version number` and put nuxt3 first

Reply:
I like option 2a. I'd like to see the order of the detectors be grouped by framework, but within the group ordered by specificity.

I also like the idea of `matchPackage` and `matchContent` be a `string | RegExp` just in case we have a use case come up where the current mechanism isn't opinionated enough.

Reply:
`matchPackage` is used to build the regex. So, it could include regex syntax. That's not clear from the type, but I'm also not passing in a regex to build another regex will be worth making this harder to debug.